### ISSUE TITLE
Changes needed for generating persistent kernel

### ISSUE THREAD

**Issue by FDecaYed** [csarofeen/pytorch#389]
Below code is used in attempt to generate persistent softmax kernel. The idea is let each block handle fixed number of element, and cache input in reg/smem. Dynamic size on reduction dim is handled with number of block and grid reduction.

## test code
```
void testGPU_FusionSoftmax2DSmem() {
  Fusion fusion;
  FusionGuard fg(&fusion);

  const int tidx = 128;
  const int elems_in_reg = 8;
  const int dimx = 4096;
  const int dimy = 1024;

  // Set up your input tensor views
  TensorView* input_tv0 = makeDummyTensor(2);
  fusion.addInput(input_tv0);
  TensorView* exp_tv1 = unaryOp(UnaryOpType::Exp, input_tv0);
  TensorView* sum_exp_tv2 = sum(exp_tv1, {-1});
  TensorView* bcast_sum_tv3 = broadcast(sum_exp_tv2, {false, true});
  TensorView* exp_tv1_copy_tv4 = unaryOp(UnaryOpType::Exp, input_tv0);
  TensorView* output_tv5 = div(exp_tv1_copy_tv4, bcast_sum_tv3);
  fusion.addOutput(output_tv5);

  sum_exp_tv2->split(-1, elems_in_reg);
  bcast_sum_tv3->split(-1, elems_in_reg);
  output_tv5->split(-1, elems_in_reg);
  sum_exp_tv2->split(-2, tidx);
  bcast_sum_tv3->split(-2, tidx);
  output_tv5->split(-2, tidx);

  TensorView* sum_exp_rf_tv6 = sum_exp_tv2->rFactor({-1});
  exp_tv1->computeAt(sum_exp_rf_tv6, -3);
  exp_tv1_copy_tv4->computeAt(output_tv5, -3);

  TensorView* cache_tv7 = exp_tv1->cache_before();
  cache_tv7->setMemoryType(MemoryType::Shared);

  TensorView* tensors_to_parallelize[] = {exp_tv1, exp_tv1_copy_tv4, cache_tv7, sum_exp_tv2, bcast_sum_tv3, output_tv5, sum_exp_rf_tv6};
  for (auto tv : tensors_to_parallelize) {
    tv->axis(0)->parallelize(ParallelType::BIDx);
    tv->axis(1)->parallelize(ParallelType::BIDy);
    tv->axis(2)->parallelize(ParallelType::TIDx);
  }
  fusion.printKernel();
  torch::jit::fuser::cuda::FusionExecutor fe;
  fe.compileFusion(&fusion);
}

```
## generated cuda code
```
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 2> T5, Tensor<float, 1> T14, Tensor<int64_t, 1> T15){
  alignas(4) extern __shared__ char array[];
  void* shared_mem = array;
  float T3[1];
  float T2[1];
  T2[ 0 ]
     = float(0);
  float T6[1];
  T6[ 0 ]
     = float(0);
  float T1[8];
  __shared__ float T7[( 128 * 8 )];
  for(size_t i297 = 0; i297 < 8; ++i297 ) {
    if ( ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i297 ) < T0.size[1] ) ) {
      T7[ ( ( threadIdx.x * 8 ) + i297 ) ]
         = expf(T0[ ( blockIdx.x * T0.stride[0] ) + ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i297 ) * T0.stride[1] ) ]);
    }
  }
  __syncthreads();
  for(size_t i305 = 0; i305 < 8; ++i305 ) {
    if ( ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i305 ) < T0.size[1] ) ) {
      T1[ i305 ]
         = T7[ ( ( threadIdx.x * 8 ) + i305 ) ];
    }
  }
  for(size_t i311 = 0; i311 < 8; ++i311 ) {
    if ( ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i311 ) < T0.size[1] ) ) {
      T6[ 0 ]
         = T6[ 0 ]
         + T1[ i311 ];
    }
  }
  float block_result;
  blockReduce< true, false, false > ( block_result, T6[ 0 ], reduction_add_float, threadIdx, blockDim, static_cast<float*>(shared_mem));
  bool T2pred;
  // Allocate global tensor float T14[( T0.size[0] * ( ceilDiv(( ceilDiv(T0.size[1], 8) ), 128) ) )];
  // Allocate global tensor int64_t T15[T0.size[0]];
  T2pred = reduction::gridReduce< false, true, false, false, true, true > ( T2[ 0 ], block_result, reduction_add_float, &T14[0], T15, static_cast<float*>(shared_mem));
  broadcast::blockBroadcast<true, false, false>(T3[ 0 ], T2[ 0 ], static_cast<float*>(shared_mem));
  // place holder for grid broadcast.
  float T4[8];
  for(size_t i336 = 0; i336 < 8; ++i336 ) {
    if ( ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i336 ) < T0.size[1] ) ) {
      T4[ i336 ]
         = expf(T0[ ( blockIdx.x * T0.stride[0] ) + ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i336 ) * T0.stride[1] ) ]);
    }
  }
  for(size_t i342 = 0; i342 < 8; ++i342 ) {
    if ( ( ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i342 ) < T0.size[1] ) ) {
      T5[ ( blockIdx.x * T5.stride[0] ) + ( ( ( ( blockIdx.y * 128 ) + threadIdx.x ) * 8 ) + i342 ) ]
         = T4[ i342 ]
         / T3[ 0 ];
    }
  }
}
```
## issues

- grid broadcast is not currently supported. Here I commented out backend assert and added place holder.
- Using same tensor in branches that have different conputeAt is not supported, thus we created tv4 as a copy of tv1 to workaround. As a result, there is no way to reuse tv1 even though it is cached in tv7, and tv4 have same loop structure.

**Comment by rdspring1** [comment]
Here is a persistent Softmax kernel that uses dynamic shared memory and does not require grid broadcast. Essentially, each block handles a row of the input.

**Issues:**
- I couldn't reuse a `TensorView` in certain `ComputeAt` cases because of tensor replication. Is it possible to loosen this restriction for reading from shared memory? My workaround was to directly read from shared memory, but it requires extra register usage.

**Schedule:**
```
void testGPU_FusionSoftmax2DSmemBlock() {
  const size_t tidx = 128;

  // Set up your input tensor views
  TensorView* input_tv0 = makeDummyTensor(2);
  fusion.addInput(input_tv0);
  TensorView* exp_tv1 = unaryOp(UnaryOpType::Exp, input_tv0); // (M, N)
  TensorView* sum_exp_tv2 = sum(exp_tv1, {-1}); // (M, R)
  TensorView* bcast_sum_tv3 = broadcast(sum_exp_tv2, {false, true}); // (M, B)

  // Read Input into Shared Memory
  TensorView* cache_tv4 = exp_tv1->cache_before();
  cache_tv4->setMemoryType(MemoryType::Shared);

  // Read from shared memory cache
  TensorView* cache_tv5 = unaryOp(UnaryOpType::Set, cache_tv4); 

  TensorView* output_tv6 = div(cache_tv5, bcast_sum_tv3); // (M, N) = (M, N) / (M, B)
  fusion.addOutput(output_tv6);

  // M, N/128, 128
  exp_tv1->split(-1, tidx);
  sum_exp_tv2->split(-1, tidx);
  bcast_sum_tv3->split(-1, tidx);
  cache_tv4->split(-1, tidx);
  cache_tv5->split(-1, tidx);
  output_tv6->split(-1, tidx);

  TensorView* sum_exp_rf_tv6 = sum_exp_tv2->rFactor({1});

  cache_tv4->computeAt(exp_tv1, 1);
  exp_tv1->computeAt(sum_exp_rf_tv6, -1);
  sum_exp_rf_tv6->computeAt(sum_exp_tv2, 1);
  cache_tv5->computeAt(output_tv6, -1);

  TensorView* tensors_to_parallelize[] = {exp_tv1, cache_tv4, sum_exp_tv2, bcast_sum_tv3, cache_tv5, output_tv6, sum_exp_rf_tv6};
  for (auto tv : tensors_to_parallelize) {
    tv->axis(0)->parallelize(ParallelType::BIDx);
    tv->axis(-1)->parallelize(ParallelType::TIDx);
  }

  fusion.printKernel();
  torch::jit::fuser::cuda::FusionExecutor fe;
  fe.compileFusion(&fusion);
}
```
**Generated Cuda Kernel:**
```
__global__ void CUDAGeneratedKernel(Tensor<float, 2> T0, Tensor<float, 2> T6) {
  alignas(4) extern __shared__ char array[];
  unsigned offset = 0;
  void* shared_mem = array;
  offset += ((blockDim.x * blockDim.y * blockDim.z) * sizeof(float));

  offset = alignBufferSize(offset,4);
  float* T4 = reinterpret_cast<float*>(array + offset);
  offset += (((ceilDiv(T0.size[1], 128)) * 128) * sizeof(float));

  float T3[1];

  float T2[1];
  T2[0] = float(0);

  float T7[1];
  T7[0] = float(0);

  // Read Input into Shared Memory
  for(size_t i4 = 0; i4 < (ceilDiv(T0.size[1], 128)); ++i4) {
    if ((((i4 * 128) + threadIdx.x) < T0.size[1])) {
      T4[((i4 * 128) + threadIdx.x)]
         = expf(T0[(blockIdx.x * T0.stride[0]) + (((i4 * 128) + threadIdx.x) * T0.stride[1])]);
    }
  }
  __syncthreads();

  // Reduction
  for(size_t i11 = 0; i11 < (ceilDiv(T0.size[1], 128)); ++i11) {
    float T1[1];
    if ((((i11 * 128) + threadIdx.x) < T0.size[1])) {
      T1[0]
         = T4[((i11 * 128) + threadIdx.x)];
    }
    if ((((i11 * 128) + threadIdx.x) < T0.size[1])) {
      T7[0]
        = T7[0]
        + T1[0];
    }
  }
  blockReduce<true, false, false>(
    T2[0],
    T7[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    1,
    float(0));

  // Block Broadcast
  broadcast::blockBroadcast<true, false, false>(
    T3[0],
    T2[0],
    static_cast<float*>(shared_mem));

  // Division + Read from shared memory to registers
  // Issue: duplicate registers
  for(size_t i32 = 0; i32 < (ceilDiv(T0.size[1], 128)); ++i32) {
    float T5[1];
    if ((((i32 * 128) + threadIdx.x) < T0.size[1])) {
      T5[0]
         = T4[((i32 * 128) + threadIdx.x)];
    }
    if ((((i32 * 128) + threadIdx.x) < T0.size[1])) {
      T6[(blockIdx.x * T6.stride[0]) + ((i32 * 128) + threadIdx.x)]
        = T5[0]
        / T3[0];
    }
  }
}
```

**Comment by csarofeen** [comment]
Closing in favor of https://github.com/csarofeen/pytorch/issues/1399
